nomad/state/state_store.go

```diff
@@ -526,7 +526,21 @@ func (s *StateStore) UpsertNode(index uint64, node *structs.Node) error {
 		node.CreateIndex = exist.CreateIndex
 		node.ModifyIndex = index
 		node.Drain = exist.Drain // Retain the drain mode
+
+		// retain node events that have already been set on the node
+		node.NodeEvents = exist.NodeEvents
 	} else {
+		// Because this is the first time the node is being registered, we should
+		// also create a node registration event
+		nodeEvent := &structs.NodeEvent{
+			Message:   "Node Registered",
+			Subsystem: "Cluster",
+			Timestamp: node.StatusUpdatedAt,
+		}
+
+		node.NodeEvents = make([]*structs.NodeEvent, 0)
```

Comment:
Make it at least size 1 so it doesn't resize immediately: `make([]*structs.NodeEvent, 0, 1)`